lib/remap-lang/src/expression/for_loop.rs

```diff
@@ -0,0 +1,129 @@
+    Collection(#[from] value::Error),
+
+    #[error("iterating over a map requires two variables, one for the key and one for the value")]
+    MissingMapVariable,
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct ForLoop {
+    collection: Box<Expr>,
+
+    // map key or array value
+    var1: Variable,
+
+    // map value or array index
+    var2: Option<Variable>,
+
+    // block
+    expression: Box<Expr>,
+}
+
+impl ForLoop {
+    pub fn new(
+        collection: Expr,
+        var1: Variable,
+        var2: Option<Variable>,
+        expression: Expr,
+        state: &state::Compiler,
+    ) -> Result<Self> {
+        let type_def = collection.type_def(state);
+
+        // We expect this to resolve to exactly an array or a map, so that we
+        // know what to assign to the first (and optional second) variable.
+        if !type_def.kind.is_array() && !type_def.kind.is_map() {
+            return Err(E::from(Error::Collection(value::Error::Expected(
+                value::Kind::Array | value::Kind::Map,
+                type_def.kind,
+            )))
+            .into());
+        }
+
+        // When dealing with a map, the second variable identifier is required.
+        if type_def.kind.is_map() && var2.is_none() {
+            return Err(E::from(Error::MissingMapVariable).into());
+        }
+
+        let collection = Box::new(collection);
+        let expression = Box::new(expression);
+
+        Ok(Self {
+            collection,
+            var1,
+            var2,
+            expression,
+        })
+    }
+}
+
+impl Expression for ForLoop {
+    fn execute(&self, state: &mut state::Program, object: &mut dyn Object) -> Result<Value> {
+        let collection = self.collection.execute(state, object)?;
+
+        match collection {
+            Value::Array(array) => {
+                array
+                    .into_iter()
+                    .enumerate()
+                    .map(|(index, value)| {
+                        // If we have a second variable, the first becomes the
+                        // index, otherwise it's the value.
+                        let var_value = match &self.var2 {
+                            Some(var2) => var2,
+                            None => &self.var1,
+                        };
+
+                        state
+                            .variables_mut()
+                            .insert(var_value.ident().to_owned(), value);
+
+                        if self.var2.is_some() {
+                            state
+                                .variables_mut()
+                                .insert(self.var1.ident().to_owned(), (index as i64).into());
+                        }
+
+                        self.expression.execute(state, object)
+                    })
+                    .collect::<Result<Vec<_>>>()?;
+
+                // purge variables after loop ends
+                state.variables_mut().remove(&self.var1.ident().to_owned());
```

Comment:
This could be dangerous, if the loop variable is shadowing an existing name.

```
x = 3
for x in [1,2] {
...
}
# x should still be around
log(x)
```

I think to implement this change we would need to to turn the variables store into some sort of stack, in order to maintain the nested state.

Reply:
I agree. I've been thinking about scoping variables. We're not too far away from supporting that, and it makes these issues go away. I'm not entirely convinced yet, but it would fit more closely with any type-level scoping we're going to add in the future:

```rust
if is_string(.foo) { 
	.bar = upcase(.foo)
}
```